sqlitecluster/SQLiteNode.cpp

```diff
@@ -74,6 +80,143 @@ SQLiteNode::~SQLiteNode() {
 }
 
+void SQLiteNode::replicate(SQLiteNode& node, SQLite& db, int threadNum) {
+    SInitialize("repl" + to_string(threadNum));
+    auto _state = node._state.load();
+    string name = "xxx";
+
+    // If we're already in the middle of handling a transaction, we'll set this to true. This allows us to re-use the
+    // main loop and wait logic.
+    // 0 = no command.
+    // 1 = command not started.
+    // 2 = command begun
+    int transactionState = 0;
+    uint64_t commandCommitCount = 0;
+
+    // Create a dummy placeholder command.
+    pair<Peer*, SData> p(nullptr, SData(""));
+
+    while (true) {
+        // If we're not already handling a transaction, let's see if we can get a new command.
+        if (!transactionState) {
+            unique_lock<mutex> lock(node._replicationCommandMutex);
+            if (node._replicationThreadsShouldExit) {
+                // Every time we lock, we may be falling out of FOLLOWING, and might not receive any more
+                // notifications, so we need to check this each time the lock is acquired.
+                return;
+            }
+            try {
+                p = node._replicationCommands.pop();
+            } catch (const out_of_range& e) {
+                // We needed a command, but we couldn't find one. Wait until something happens, and then start our
+                // loop over.
+                node._replicationCV.wait(lock);
+                continue;
+            }
+
+            // We just got a new command.
+            auto& command = p.second;
+            if (SIEquals(command.methodLine, "BEGIN_TRANSACTION")) {
+                commandCommitCount = command.calcU64("NewCount");
+                // Indicate that this command is starting.
+                transactionState = 1;
+            }
+            if (SIEquals(command.methodLine, "ROLLBACK_TRANSACTION")) {
+                lock_guard<mutex> hashlock(node._replicationHashMutex);
+                node._rollbackHashes.insert(command["NewHash"]);
+                node._replicationCV.notify_all();
+                continue;
+            }
+            if (SIEquals(command.methodLine, "COMMIT_TRANSACTION")) {
+                commandCommitCount = command.calcU64("CommitCount");
+                lock_guard<mutex> hashlock(node._replicationHashMutex);
+                node._replicationHashes.emplace(make_pair(commandCommitCount, command["Hash"]));
+                node._replicationCV.notify_all();
+                continue;
+            }
+        }
+
+        try {
+            // Convenience vars.
+            auto& peer = p.first;
+            auto& command = p.second;
+
+            // The following block only runs if there's already a command ready to run. This includes a command we just
+            // created
+            if (transactionState == 1) {
+                unique_lock<mutex> lock(node._replicationCommandMutex);
+                if (node._replicationThreadsShouldExit) {
+                    return;
```

Comment:
A hmm would be good here because we didn't finish processing this

Reply:
This is allowed, we're not obligated to complete replication before shutting down.

Reply:
I don't think it's even a `hmm` though, I think it's just noise, it'll happen every time we switch away from FOLLOWING.